# Virseqimprover

An integrated pipeline for error-correction, extension, and annotation of viral contigs to recover error-free full genomes of viruses and phages by polishing and extending draft viral assemblies.

# Installation

Use conda install to install Virseqimprover package.

```bash
conda install Virseqimprover --channel haoqiusong
```
This will help you install the dependent tools (e.g. SPAdes, Bowtie2, etc.) and solve potential dependency problems you might have.

# Run Virseqimprover

## Input

A viral contig (.fasta) and the read sample(s) (.fastq) from which the input contig is assembled.

The reads can be paired-end reads or single-end reads. For paired-end reads, two read files are needed.

## Usage

Command line example:
```bash
python Virseqimprover.py
-1 /home/Virseqimprover/input/read1.fastq (-2 /home/Virseqimprover/input/read2.fastq) -scaffold /home/Virseqimprover/input/scaffold.fasta -o /home/Virseqimprover/output
```
Using full paths for all the files and directories mentioned in the input are highly recommended.

## Output

The ouput folders and files will be generated in the output directory folder you specify. (If not specified, it will be the same directory of your input files by default.)

The final output file -- ```pilon_out.fasta``` is the final improved contig of the original contig.

# Annotation example

1. After applying Virseqimprover to the viral contig, use an online tool eggNOG-Mapper (http://eggnog-mapper.embl.de/) to annotate viral genes. On the website, choose "Metagenomic" as the data kind, Prodigal as the gene prediction method, then upload the improved contig file, and choose "Viruses - 10239" as the taxonomic scope.

2. When the job is done, download the ```out.emapper.decorated.gff``` and ```out.emapper.annotations``` files, use them to generate two separated annotation tables (.csv), including the annotated and unannotated genes by running ```ann.py```.

Command line example:
```bash
python ann.py
./out.emapper.annotations
./out.emapper.decorated.gff
```

3. Upload ```pilon_out.fasta``` (as the sequence) and the two .csv files (as the features) generated by ```ann.py``` to an online visualization tool Proksee (https://proksee.ca/) to visualize the annotation of the improved viral contig.

# Possible problems and solutions

1. In case conda can not solve all the dependency issues, try to create a new environment and install again.
Command lines:
```bash
conda create -n <new_environment_name>
conda activate <new_environment_name>
conda install Virseqimprover --channel haoqiusong
```

2. In case you meet a problem for Samtools ("samtools: error while loading shared libraries: libcrypto.so.1.0.0: cannot open shared object file: No such file or directory"), use the following command lines to solve it:

```bash
cd anaconda3/envs/<environment_name>/lib
ll libcrypto.so (Then you might see "libcrypto.so -> libcrypto.so.1.1")
ln -s libcrypto.so.1.1 libcrypto.so.1.0.0
```

3. In case you do not have the permission to run the shell file, use command ```chmod +x ***.sh```.
